docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,206 @@
+- `name`: A short string identifying the new version.
+- `height`: The chain height at which the upgrade is to be performed.
+- `info`: A string containing information about the upgrade.
+
+The `info` string can be anything.
+However, Cosmovisor will try to use the `info` field to automatically download a new version of the blockchain executable.
+For the auto-download to work, Cosmovisor expects it to be either a stringified JSON object (with a specific structure defined through documentation), or a URL that will return such JSON.
+The JSON object identifies URLs used to download the new blockchain executable for different platforms (OS and Architecture, e.g. "linux/amd64").
+Such a URL can either return the executable file directly or can return an archive containing the executable and possibly other assets.
+
+If the URL returns an archive, it is decompressed into `{DAEMON_HOME}/cosmovisor/{upgrade name}`.
+Then, if `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}` does not exist, but `{DAEMON_HOME}/cosmovisor/{upgrade name}/{DAEMON_NAME}` does, the latter is copied to the former.
+If the URL returns something other than an archive, it is downloaded to `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}`.
+
+Both `DAEMON_HOME` and `DAEMON_NAME` are [environment variables used to configure Cosmovisor](https://github.com/cosmos/cosmos-sdk/blob/cosmovisor/v1.0.0/cosmovisor/README.md#command-line-arguments-and-environment-variables).
+
+Currently, there is no mechanism that makes Cosmovisor run a command after the upgraded chain has been restarted.
+
+## Decision
+
+### Protobuf Updates
+
+We will define a new message for providing upgrade instructions and add it as a new field in the `Plan` message.
+The upgrade instructions will contain a list of artifacts available for each platform.
+It will also allow for the definition of a pre-run and post-run command.
+
+```protobuf
+message Plan {
+  // ... (existing fields)
+
+  UpgradeInstructions upgrade = 6;
+}
+```
+
+The new `UpgradeInstructions upgrade` field MUST be optional.
+
+```protobuf
+message UpgradeInstructions {
+  string pre_run              = 1;
+  string post_run             = 2;
+  repeated Artifact artifacts = 3;
+  string description          = 4;
+}
+```
+
+All fields in the `UpgradeInstructions` are optional.
+- `pre_run` is a command to run prior to the upgraded chain restarting.
+  If defined, then the app supervisors (e.g. Cosmovisor) MUST NOT run `app pre-run`.
+  If defined, this command MUST behave the same as the current [pre-upgrade](https://github.com/cosmos/cosmos-sdk/blob/v0.44.5/docs/migrations/pre-upgrade.md) command.
+  The working directory this command runs from MUST be `{DAEMON_HOME}/cosmovisor/{upgrade name}`.
+- `post_run` is a command to run after the upgraded chain has been started. If defined, this command MUST be only executed once.
```

Comment:
how do we expect to be able to enforce only-once semantics?

Reply:
I updated a few spots on this to clarify that it's once per node (not once per chain).